### ISSUE TITLE
I think the train process have some trouble!!!

### ISSUE THREAD

**Issue by Johnson-yue** [znxlwm/UGATIT-pytorch#17]
Hi, thank you for sharing your code with pytorch, and I find some importance trouble in your code. Maybe cause failed to reimplements model performance!!

In UGATIT.py -- train function:
 1) when you update D  , you do not use

> self.genA2B.eval(), self.genB2A.eval() and in [your code here](https://github.com/znxlwm/UGATIT-pytorch/blob/bc293fa3afe7585157d5542d135340cb3954bede/UGATIT.py#L167)

> fake_B2A.detach(), fake_B2A.detach() ..... but only [fake_B2A](https://github.com/znxlwm/UGATIT-pytorch/blob/bc293fa3afe7585157d5542d135340cb3954bede/UGATIT.py#L210),  [fake_B2A](https://github.com/znxlwm/UGATIT-pytorch/blob/bc293fa3afe7585157d5542d135340cb3954bede/UGATIT.py#L211).   and so on...

so, when you want to only update D, but meanwhile, the G is also updated.
It is the same error in **update G** 

you can reference [SPADE](https://github.com/NVlabs/SPADE/blob/929194ef05c2c523c8fc8729f140059c40064684/models/pix2pix_model.py#L170) code , if I am wrong ,please tell me

**Comment by phongnhhn92** [comment]
Hi, 
For the 1st comment, I think that it is unnecessary to set both `self.genA2B.eval(), self.genB2A.eval()` into eval mode because we are in the training function.
On the 2nd comment, I think it is also unnecessary to detach both fake_B2A, fake_B2A before interesting them into the Discriminator. The idea of detach() is constructing a new view on a tensor which is declared not to need gradients. However, as you mentioned in the SPADE code, right after detach() then the next line is             fake_image.requires_grad_() which reverse above line of code. It is kind a weird why SPADE code is like that but so far, there is no problem for me.

**Comment by Johnson-yue** [comment]
Ok, it also work for me. I think I will check when use the detach() function in GAN models. Thank you

**Comment by shoutOutYangJie** [comment]
> Ok, it also work for me. I think I will check when use the detach() function in GAN models. Thank you

Can you tell me how better you reproduce this project? I find the author dosen't release pytorch weight. I can't ensure this project is correct compared to the official tensorflow implementition.
